### ISSUE TITLE
S3 persistence may experience a problematic delay

### ISSUE THREAD

**Issue by mdb** [GREsau/localstack-persist#8]
👋 Hi - thanks for your work on `localstack-persist`. If I understand correctly, I believe there may be a bug wherein there's a problematic delay (or, perhaps a failure altogether?) persisting data to the `/persisted-data` directory (or perhaps a delay persisting data to the host machine directory mounted at `/persisted-data`?).

My work at [mdb/tf-workspaces-demo](https://github.com/mdb/tf-workspaces-demo) may have inadvertently uncovered this problem (see the [docker-compose.yaml](https://github.com/mdb/tf-workspaces-demo/blob/main/docker-compose.yaml) for details on my `localstack-persist`'s usage). For context, [mdb/tf-workspaces-demo](https://github.com/mdb/tf-workspaces-demo) uses `localstack-persist` to share `localstack-persist` persisted data between GitHub Actions jobs; in this case, the GH Actions workflow uses [actions/upload-artifact](https://github.com/actions/upload-artifact) to upload the data such that a downstream job can can download it via [actions/download-artifact](https://github.com/actions/download-artifact) and run its own `localstack-persist` container using this data. 

After creating an S3 object, I've discovered it's necessary to [poll /localstack-persist/s3/store.json for the expected S3 bucket contents](https://github.com/mdb/tf-workspaces-demo/blob/main/scripts/wait-for-localstack-persistence.sh), even after an HTTP GET (from the host machine) to the specific S3 object HTTP endpoint has succeeded with HTTP response code 200. The object's key is not immediately present in `/localstack-persist/s3/store.json`.

After updating an S3 object, I've also discovered that `/localstack-persist/s3/store.json` may report an incorrect, stale `size` attribute associated with an S3 object for a period of time, causing HTTP requests against the S3 object following a `localstack-persist` container restart to return an HTTP response with an incorrect `Content-Length` header. For example...

The `/persisted-data/s3/assets/tf-workspaces-demo/env%3a%2f456_us-east-1_dev_pr-15%2fterraform.tfstate@null` file is clearly 2219 bytes:

```
$ docker compose exec -T localstack-persist ls -l /persisted-data/s3/assets/tf-workspaces-demo
total 4
-rw-r--r-- 1 root root 2219 Dec 20 12:20 env%3a%2f456_us-east-1_dev_pr-15%2fterraform.tfstate@null
```

However, `localstack-persist` attempts to serve it with a `Content-Length` header specifying its length as 180 bytes, which causes the HTTP request to fail:

```
$ curl --verbose  http://localhost.localstack.cloud:4566/tf-workspaces-demo/env:/456_us-east-1_dev_pr-15/terraform.tfstate
*   Trying 127.0.0.1:4566...
* Connected to localhost.localstack.cloud (127.0.0.1) port 4566
> GET /tf-workspaces-demo/env:/456_us-east-1_dev_pr-15/terraform.tfstate HTTP/1.1
> Host: localhost.localstack.cloud:4566
> User-Agent: curl/8.4.0
> Accept: */*
>
< HTTP/1.1 200
< Content-Type: application/json
< accept-ranges: bytes
< Last-Modified: Wed, 20 Dec 2023 11:54:09 GMT
< Content-Length: 180
< ETag: "59c685d72649c72c62602143867a2bf8"
< x-amz-server-side-encryption: AES256
< x-amz-request-id: 254219a5-7dba-43b9-82db-fdb1f620b163
< x-amz-id-2: s9lzHYrFp76ZVxRcpX9+5cjAnEH2ROuNkd2BHfIa6UkFVdtjf5mKR3/eTPFvsiP/XV/VLi31234=
< Connection: close
< date: Thu, 21 Dec 2023 16:34:20 GMT
< server: hypercorn-h11
<
* transfer closed with 180 bytes remaining to read
* Closing connection
curl: (18) transfer closed with 180 bytes remaining to read
```

And `store.json` incorrectly reports its size as 180:

```
$ TERRAFORM_VERSION=1.5.5 docker compose exec -T localstack cat /persisted-data/s3/store.json | jq -r '.data["000000000000"]["us-east-1"]["_global"]["buckets"]["tf-workspaces-demo"]["objects"]["_store"]'
{
  "env:/456_us-east-1_dev_pr-15/terraform.tfstate": {
    "py/object": "localstack.services.s3.v3.models.S3Object",
    "key": "env:/456_us-east-1_dev_pr-15/terraform.tfstate",
    "user_metadata": {},
    "system_metadata": {
      "ContentType": "application/json"
    },
    "version_id": null,
    "storage_class": "STANDARD",
    "etag": "59c685d72649c72c62602143867a2bf8",
    "size": 180,
    "expires": null,
    "checksum_algorithm": null,
    "checksum_value": null,
    "encryption": "AES256",
    "kms_key_id": null,
    "bucket_key_enabled": false,
    "lock_mode": null,
    "lock_legal_status": null,
    "lock_until": null,
    "acl": {
      "Owner": {
        "DisplayName": "webfile",
        "ID": "75aa57f09aa0c8caeab4f8c24e99d10f8e7faeebf76c078efc7c6caea54ba06a"
      },
      "Grants": [
        {
          "Grantee": {
            "DisplayName": "webfile",
            "ID": "75aa57f09aa0c8caeab4f8c24e99d10f8e7faeebf76c078efc7c6caea54ba06a",
            "Type": "CanonicalUser"
          },
          "Permission": "FULL_CONTROL"
        }
      ]
    },
    "expiration": null,
    "website_redirect_location": null,
    "is_current": true,
    "last_modified": {
      "py/object": "datetime.datetime",
      "isoformat": "2023-12-20T11:54:09.152952+00:00"
    },
    "parts": {},
    "restore": null,
    "owner": {
      "py/id": 12
    }
  }
}
```

**Comment by GREsau** [comment]
Changes may indeed take a few seconds to actually be written to the `/persisted-data` volume, and that's by-design to avoid frequent requests causing many writes to disk. When a resource is created/updated/changed in a running LocalStack container, the change is applied to an in-memory data-store That in-memory data is the "source of truth" for most resource state, and is what's used to set the object metadata in S3 responses. This data-store is then written to disk on a schedule (currently every 10 seconds, although this is subject to change) and also on container shutdown. 

I would recommend against probing the contents of the `/persisted-data` directory, as both the directory structure and file contents are implementation details of localstack-persist, and so are also liable to change, e.g. it's likely a future version will serialise data as a binary format instead of JSON.

Having said that, the S3 GET returning the wrong Content-Length does look like a bug! Is it reproducible? I cloned mdb/tf-workspaces-demo and ran `docker compose up` and didn't see any errors occur, but presumably that's not enough to trigger the bug?

**Comment by mdb** [comment]
> and also on container shutdown.

Thanks for the quick response! In in my case, I wasn't shutting down the container before attempting to use [actions/upload-artifact](https://github.com/actions/upload-artifact) to upload the persisted data; I'd overlooked that doing so would suffice in persisting data, circumventing the need to probe `/persisted-data` directory contents. So, that's actually a great solution for me. See https://github.com/mdb/tf-workspaces-demo/pull/22 for the revised GH Actions workflow, if you're curious.

> the S3 GET returning the wrong Content-Length does look like a bug!

I'm fairly confident this is just the result of the same by-design delay in writing data to disk you've explained. Again, in my case, I wasn't shutting down the container before attempting to use [actions/upload-artifact](https://github.com/actions/upload-artifact) to upload the persisted data and pass it along to a downstream GH Actions build. As a result, the downstream build attempted to use a http://localhost.localstack.cloud:4566/tf-workspaces-demo/env:/456_us-east-1_dev_pr-15/terraform.tfstate object whose entry in `/persisted-data/s3/store.json` didn't yet reflect the correct `"size"` (and therefore the `Content-Length` returned in HTTP responses was wrong).

So, in summary, it sounds like `localstack-persist` is working by design and I fell victim to a gotcha I hadn't considered. Thanks again @GREsau !
